applications/DG-Max/Algorithms/DGMaxEigenvalue.cpp

```diff
@@ -671,29 +635,24 @@ void DGMaxEigenvalue<DIM>::SolverWorkspace::initStiffnessMatrixShifts() {
 
 template <std::size_t DIM>
 void DGMaxEigenvalue<DIM>::SolverWorkspace::extractEigenVectors() {
+    std::swap(eigenpairs_, previousEigenpairs_);
     PetscErrorCode error;
-    error = EPSGetConverged(solver_, &convergedEigenValues_);
-    CHKERRABORT(PETSC_COMM_WORLD, error);
-    if (convergedEigenValues_ > numberOfEigenVectors_) {
-        // Reallocate
-        DGMaxLogger(INFO,
-                    "Reallocating eigenvector storage to store % instead of % "
-                    "eigenvectors",
-                    convergedEigenValues_, numberOfEigenVectors_);
-        error = VecDestroyVecs(numberOfEigenVectors_, &eigenVectors_);
-        CHKERRABORT(PETSC_COMM_WORLD, error);
-        error = VecDuplicateVecs(tempFieldVector_, convergedEigenValues_,
-                                 &eigenVectors_);
-        CHKERRABORT(PETSC_COMM_WORLD, error);
-        numberOfEigenVectors_ = convergedEigenValues_;
-    }
-
-    eigenvalues_.resize(convergedEigenValues_);
-    for (PetscInt i = 0; i < convergedEigenValues_; ++i) {
-        EPSGetEigenpair(solver_, i, &eigenvalues_[i], nullptr, eigenVectors_[i],
-                        nullptr);
-    }
+    eigenpairs_.loadEigenpairs(solver_, tempFieldVector_);
     CHKERRABORT(PETSC_COMM_WORLD, error);
+    // Reorder
+    std::vector<std::size_t> ordering(eigenpairs_.size());
+    std::iota(ordering.begin(), ordering.end(), 0);
+    std::sort(ordering.begin(), ordering.end(),
+              [&](const std::size_t& i1, const std::size_t& i2) {
+                  PetscScalar e1 = eigenpairs_.getEigenvalue(ordering[i1]);
+                  PetscScalar e2 = eigenpairs_.getEigenvalue(ordering[i2]);
+                  if (PetscRealPart(e1) != PetscRealPart(e2)) {
```

Comment:
this line still worries me a bit, because of floating point, but I guess if petsc solves for complex pairs, the real parts are exactly equal.

Reply:
I don't see any problem with it. The goal is to provide an order on the complex numbers so that the result will be sorted and preferably in a reproducible fashion. This just defines a total order that orders the values by real part, and to make it a total order it sorts it based on imaginary as backup.